### ISSUE TITLE
--prefer-source doesn't download the source if the package has already been downloaded.

### ISSUE THREAD

**Issue by lukearmstrong** [composer/composer#2943]
_I noticed this during my first attempt at creating a composer package, so it may be accepted behaviour rather than a bug, but I will just leave this suggestion here in case others agree._

I added my new package to the `composer.json` of my application, and ran `composer update` to get it.

As I am still developing the package, I wanted to make changes to it, so read the docs and ran `composer update user-name/package-name --prefer-source` hoping that it would download the .git repository too.

```
Loading composer repositories with package information
Updating dependencies (including require-dev)
Nothing to install or update
Generating autoload files
Generating optimized class loader
```

Odd.

So then I ran `composer install user-name/package-name --prefer-source`, but that just threw an error.

```
Invalid argument user-name/package-name. Use "composer require user-name/package-name" instead to add packages to your composer.json.
```

In the end I deleted the `vendor/user-name/package-name/` folder from my application, and ran `composer update user-name/package-name --prefer-source` again.

This time it worked as expected and downloaded the `.git/` repository so I am able to go into that folder to develop the package directly in my application.

I realise the language of `--prefer-source` does not imply that you are forcing it to download the source, but even so, I would have preferred it to download the source even though it already had downloaded the package.

**Comment by Seldaek** [comment]
Closing as duplicate of #2901

**Comment by benjifisher** [comment]
It looks as though #2901 was closed as a duplicate of #3112.
